src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs

```diff
@@ -436,13 +436,30 @@ private Utf8String GetPrefixMangledSignatureName(IPrefixMangledSignature prefixM
                 sigRetTypeName = sigRetTypeName.Replace("::", "_");
             sb.Append(sigRetTypeName);
 
+            EmbeddedSignatureData[] embeddedSignatureData = signature.GetEmbeddedSignatureData();
             for (int i = 0; i < signature.Length; i++)
             {
                 sb.Append("__");
+                EmbeddedSignatureData? embeddedSignature = null;
+                if (embeddedSignatureData != null)
+                {
+                    for (int j = 0; j < embeddedSignatureData.Length; j++)
+                    {
+                        if (embeddedSignatureData[j].index == "0.1." + (i + 2) + ".1")
+                        {
+                            embeddedSignature = embeddedSignatureData[j];
+                        }
+                    }
+                }
+
                 string sigArgName = GetMangledTypeName(signature[i]);
                 if (_mangleForCplusPlus)
                     sigArgName = sigArgName.Replace("::", "_");
                 sb.Append(sigArgName);
+                if (embeddedSignature != null)
+                {
+                    sb.Append(EnterNameScopeSequence).Append(embeddedSignature.Value.type.GetDisplayNameWithoutNamespace()).Append(ExitNameScopeSequence);
```

Comment:
`GetDisplayNameWithoutNamespace` is a display name - it's not unique. We'd want `GetMangledTypeName`.

This name will not be stable because if there's modopts we don't care about, they'll still become part of the mangled name here. The compiler is multithreaded so we'll probably end up with randomly chosen signatures that are the same as far as we care, but mangle slightly differently.

I think the problem is that the type system lacks the right primitives do deal with these. Having to go back to reading modopts everywhere is making things really confusing.

I'm taking a stab at introducing a type system primitive for this so that we don't have to go to modopts all the time and then worry about whether we should include things from the modopt or not.

I'll make the existing code use this and this pull request can then use it too.